### PR TITLE
Add GRU Operator

### LINKED ISSUE #5254: GRU Operator

### PR DESCRIPTION
**Pull request by guoshengCS**
Resolves #5254
Rewrite [GatedRecurrentLayer ](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/GatedRecurrentLayer.cpp) in the new framework. Some code styles haven't formatted.

### CODE REVIEW

File: paddle/operators/math/gru_compute.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+//   } else if (type == "relu") {
+//     return HL_ACTIVATION_RELU;
+//   } else if (type == "tanh") {
+//     return HL_ACTIVATION_TANH;
+//   } else if (type == "linear" || type == "") {
+//     return HL_ACTIVATION_LINEAR;
+//   } else {
+//     PADDLE_THROW("Do not support activation type.");
+//   }
+// }
```
Comment (qingqing01) [inline paddle/operators/math/gru_compute.h]: remove these lines.

File: python/paddle/v2/framework/tests/test_gru_op.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+
+def tanh(x):
+    y = -2. * x
+    y[y > EXP_MAX_INPUT] = EXP_MAX_INPUT
+    return (2. / (1. + np.exp(y))) - 1.
+
+
+def relu(x):
+    return np.maximum(x, 0)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_gru_op.py]: Can these functions be imported from test_lstm_op.py ?

File: python/paddle/v2/framework/tests/test_gru_op.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            start_idx = end_idx
+            end_idx = start_idx + len(idx_in_seq_list[batch_idx])
+            batch_gate[start_idx:end_idx] = g
+            batch_reset_hidden_prev[start_idx:end_idx] = r_h_p
+            batch_hidden[start_idx:end_idx] = h
+            hidden[idx_in_seq_list[batch_idx]] = h
+        return batch_gate, batch_reset_hidden_prev, hidden
+
+    def set_data(self):
+        lod = [[0, 2, 6, 9]]
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_gru_op.py]: lod = [[0, 2, 6, 9]] -> lod = [[0, 2, 6, batch_size]] ?

File: paddle/operators/gru_op.cc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                        "The shape of Bias must be [1, frame_size * 3].");
+      PADDLE_ENFORCE_EQ(bias_width, frame_size * 3,
+                        "The shape of Bias must be [1, frame_size * 3].");
+    }
+    ctx->SetOutputDim("BatchGate", input_dims);
+    ctx->SetOutputDim("BatchResetHiddenPrev", {input_dims[0], frame_size});
+    ctx->SetOutputDim("BatchHidden", {input_dims[0], frame_size});
+    ctx->SetOutputDim("Hidden", {input_dims[0], frame_size});
+    // ctx->ShareLoD("Input", "Gate");
+    // ctx->ShareLoD("Input", "ResetHiddenPrev");
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/gru_op.cc]: Done.

File: paddle/operators/gru_op.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    ctx->ShareLoD("Input", "Hidden");
+  }
+};
+
+class GRUOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  GRUOpMaker(framework::OpProto* proto, framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Input",
+             "(LoDTensor) The first input is a LodTensor, which support "
```
[the remark this answers is not in the document]
Comment (guoshengCS) [reply paddle/operators/gru_op.cc]: Done.

File: paddle/operators/math/gru_compute.h (hunk 1 above)
Comment (guoshengCS) [reply paddle/operators/math/gru_compute.h]: Done.

File: python/paddle/v2/framework/tests/test_gru_op.py (hunk 2 above)
Comment (guoshengCS) [reply python/paddle/v2/framework/tests/test_gru_op.py]: Done.

File: python/paddle/v2/framework/tests/test_gru_op.py (hunk 3 above)
Comment (guoshengCS) [reply python/paddle/v2/framework/tests/test_gru_op.py]: Done.

**Pull request merged**